Add a kyverno jp command to test jmespath expressions

samj1912 opened the pull request:
Signed-off-by: Sambhav Kothari <sambhavs.email@gmail.com>

Fixes #3167 

<!--
Please link the GitHub issue this pull request resolves in the format of `#1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->
/milestone 1.6.0
## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->
/kind feature
## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->
Adds a `kyverno jp` command to test JMESPath expressions with Kyverno grammar.

### Proof Manifests

```
$ ./cmd/cli/kubectl-kyverno/kyverno jp --help
Provides a command-line interface to JMESPath, enhanced with Kyverno specific custom functions

Usage:
  kyverno jp [flags]

Flags:
      --ast                Only print the AST of the parsed expression.  Do not rely on this output, only useful for debugging purposes
  -c, --compact            Produce compact JSON output that omits nonessential whitespace
  -e, --expr-file string   Read JMESPath expression from the specified file
  -f, --filename string    Read input JSON from a file instead of stdin
  -h, --help               help for jp
  -l, --list-functions     Output a list of custom JMESPath functions in Kyverno
  -u, --unquoted           If the final result is a string, it will be printed without quotes

Global Flags:
      --add_dir_header           If true, adds the file directory to the header of the log messages
      --log_file string          If non-empty, use this log file
      --log_file_max_size uint   Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --one_output               If true, only write logs to their native severity level (vs also writing to each lower severity level)
      --skip_headers             If true, avoid header prefixes in the log messages
      --skip_log_headers         If true, avoid headers when opening log files
  -v, --v Level                  number for the log level verbosity
$ echo '{"test": "TEST"}' | ./cmd/cli/kubectl-kyverno/kyverno jp -u 'to_lower(test)'
test

$ ./cmd/cli/kubectl-kyverno/kyverno jp --ast 'to_lower(test)'

ASTFunctionExpression {
  value: "to_lower"
  children: {
    ASTField {
      value: "test"
    }
}
```


## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [x] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is: https://github.com/kyverno/website/issues/463
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

chipzoller commented:
Do you think it would be possible (and valuable) to have a flag which actually shows the Kyverno-specific JMESPath filters that the command is capable of testing?

samj1912 commented:
> Do you think it would be possible (and valuable) to have a flag which actually shows the Kyverno-specific JMESPath filters that the command is capable of testing?

Should be possible to enumerate them yes. How would you like them to be listed?

chipzoller commented:
I was thinking of a flag that could be supplied which would output a list of them, the inputs, their types, and the output type. The problem this solves is
1. having to go to the documentation each time and
2. hoping we've updated the documentation by then :)

samj1912 commented:
@chipzoller 

```
./cmd/cli/kubectl-kyverno/kyverno jp -l
add(any,any) any
base64_decode(string) string
base64_encode(string) string
compare(string,string) bool
divide(any,any) any (divisor must be non zero)
equal_fold(string,string) bool
label_match(object,object) bool (object arguments must be enclosed in backticks; ex. `{{request.object.spec.template.metadata.labels}}`)
modulo(any,any) any (divisor must be non-zero, arguments must be integers)
multiply(any,any) any
parse_json(string) any (decodes a valid JSON encoded string to the appropriate type. Opposite of `to_string` function)
parse_yaml(string) any (decodes a valid YAML encoded string to the appropriate type provided it can be represented as JSON)
path_canonicalize(string) string
pattern_match(string,string|number) bool ('*' matches zero or more alphanumeric characters, '?' matches a single alphanumeric character)
regex_match(string,string|number)
regex_replace_all(string,string|number,string|number) string (converts all parameters to string)
regex_replace_all_literal(string,string|number,string|number) string (converts all parameters to string)
replace(string,string,string,number) string
replace_all(string,string,string) string
semver_compare(string,string) bool
split(string,string) array[string]
subtract(any,any) any
time_since(string,string,string) string
to_lower(string) string
to_upper(string) string
trim(string,string) string
truncate(string,number) string (length argument must be enclosed in backticks; ex. "{{request.object.metadata.name | truncate(@, `9`)}}")
```

That should be better. I have updated the current output to match what is available at https://main.kyverno.io/docs/writing-policies/variables/#jmespath-custom-functions with some slight updates to types to closely represent what was actually being used.

samj1912 commented:
@JimBugwadia this should be ready for review :)

JimBugwadia commented:
Hi @samj1912 - do we need this in 1.6.0? Since we are close to the GA, lets limit 1.6.0 to critical fixes. We can target the next patch or minor release. Thanks!

chipzoller commented:
> Hi @samj1912 - do we need this in 1.6.0? Since we are close to the GA, lets limit 1.6.0 to critical fixes. We can target the next patch or minor release. Thanks!

I think we need to decide whether this makes 1.6.0 despite it not being a critical fix or wait and introduce it in 1.7.0. Our goal from this point forward should be, in patch releases, to not introduce any new functionality and bug/security fixes only. If this hits 1.6.1 we violate that, which has a downstream ripple effect such as now having to maintain docs for 1.6.0 and 1.6.1. That's not something I look forward to doing, speaking personally. So what's it gonna be here?

samj1912 commented:
It depends on how much we care about this feature and the risk it poses.

From a purely risk point of view, this new feature is fairly low risk in terms of introducing issues as it is a CLI only feature and is pretty much standalone.

As for whether or not this risk is justified to include it in 1.6 for the feature it offers is upto all of you. Since this is just a CLI update, it's also fairly trivial to install it via `go install` from the main branch even if it isn't released directly, so I don't have any strong opinions.

chipzoller commented:
My vote is to include it in 1.6.0 as the risk is isolated and doesn't impact the work of the webhook. Even from a CLI blast radius standpoint, the risk is isolated. The potential for problem is due to the fact that we still have gaps in the CLI's ability to act faithfully in regression tests. The utility of this `jp` sub-command is of such value that it's hard to ignore and would dovetail nicely with the new section on JMESPath that I'm planning on writing for the 1.6.0 docs.

JimBugwadia commented:
Shuting released `1.6.0-rc3`, and was planning on releasing `1.6.0` on Monday.

If we want to include this in `1.6.0` we will need to create another RC, test, and release 1.6.0 on Tue (unless something else comes up). 

I am good with that, but we really need to get 1.6.0 completed, and there always seems to be "one last thing" :-)! 

@realshuting - what do you think?

chipzoller commented:
@samj1912 I see some filters that are missing from that output above. Is it that list statically maintained? If so, we need to update it. I'm going through the kyverno/website changes for 1.6.0 this weekend and updating docs to add missing filters, but we should also scour the code to make sure none are getting missed.

samj1912 commented:
@chipzoller, nope it's dynamically generated from our registered filters. It will only list the ones that were added specifically in kyverno.

chipzoller commented:
I withdraw my earlier comment. All the filters appear to be represented.

realshuting commented:
I have closed milestone 1.6.0 last Fri, re-opening it to include this fix.

Let's do another RC for 1.6 and we can release the official 1.6.0 the next day.

realshuting commented:
It would also be nice to standardize how to approach the releases. Some of the best practices we can follow:

- plan for releases early, and update milestone issues accordingly
- set a freeze / cutoff date for all features and enhancements
- keep posting release status during release candidates (this way we can have everyone on the same page, and communicate more efficiently to include last-minute fixes/enhancements)

Please feel free to add more and we can discuss this in the next contributors meeting.

Agile Software Development for your interest:
- https://www.agilealliance.org/agile101/
- https://www.atlassian.com/agile#:~:text=Agile%20is%20an%20iterative%20approach,faster%20and%20with%20fewer%20headaches.&text=Requirements%2C%20plans%2C%20and%20results%20are,for%20responding%20to%20change%20quickly.

chipzoller commented:
> I have closed milestone 1.6.0 last Fri, re-opening it to include this fix.
> 
> Let's do another RC for 1.6 and we can release the official 1.6.0 the next day.

I don't think one day allows sufficient time for any RC to be tested properly in the community. One of the standards we need to align on is that interval between RC and release.

JimBugwadia commented:
Hi @chipzoller - I agree we should plan for 1-2 weeks for the overall release candidate to GA process. For 1.6.0, RC1 was created 13 days ago. This update (RC4) contains only requested CLI enhancement, hence the quicker cycle.  Lets discuss the release processes in our next contributors meeting, so we can plan accordingly for future releases.
